arwen/host/execution_async.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	currentCallGroup := asyncContext.AsyncCallGroups[currentGroupID]
+	currentCallGroup.DeleteAsyncCall(asyncCallIndex)
+
+	if currentCallGroup.HasPendingCalls() {
+		return nil
+	}
+
+	asyncContext.DeleteAsyncCallGroup(currentGroupID)
+
+	// Are we still waiting for callbacks to return?
+	if asyncContext.HasPendingCallGroups() {
+		return nil
+	}
+
+	err = host.deleteCurrentAsyncContext()
+	if err != nil {
+		return err
+	}
+
+	return host.executeAsyncContextCallback(asyncContext)
+}
+
+// executeAsyncContextCallback will either execute a sync call (in-shard) to
+// the original caller by invoking its callback directly, or will dispatch a
+// cross-shard callback to it.
+func (host *vmHost) executeAsyncContextCallback(asyncContext *arwen.AsyncContext) error {
+	execMode, err := host.determineExecutionMode(asyncContext.CallerAddr, asyncContext.ReturnData)
+	if err != nil {
+		return err
+	}
+
+	if execMode != arwen.SyncExecution {
+		return host.sendContextCallbackToOriginalCaller(asyncContext)
+	}
+
+	// The caller is in the same shard, execute its callback
+	callbackCallInput := host.createSyncContextCallbackInput(asyncContext)
+
+	callbackVMOutput, callBackErr := host.ExecuteOnDestContext(callbackCallInput)
+	host.finishSyncExecution(callbackVMOutput, callBackErr)
+
+	return nil
+}
+
+func (host *vmHost) createSyncContextCallbackInput(asyncContext *arwen.AsyncContext) *vmcommon.ContractCallInput {
+	runtime := host.Runtime()
+	metering := host.Metering()
+
+	argParser := parsers.NewCallArgsParser()
+	_, arguments, err := argParser.ParseData(string(asyncContext.ReturnData))
+	if err != nil {
+		arguments = [][]byte{asyncContext.ReturnData}
+	}
+
+	input := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:     runtime.GetSCAddress(),
+			Arguments:      arguments,
+			CallValue:      runtime.GetVMInput().CallValue,
+			CallType:       vmcommon.AsynchronousCallBack,
+			GasPrice:       runtime.GetVMInput().GasPrice,
+			GasProvided:    metering.GasLeft(),
+			CurrentTxHash:  runtime.GetCurrentTxHash(),
+			OriginalTxHash: runtime.GetOriginalTxHash(),
+		},
+		RecipientAddr: asyncContext.CallerAddr,
+		Function:      arwen.CallbackFunctionName, // TODO currently default; will customize in AsynContext
+	}
+	return input
+}
+
+func (host *vmHost) loadCurrentAsyncContext() (*arwen.AsyncContext, error) {
+	runtime := host.Runtime()
+	storage := host.Storage()
+
+	asyncContext := &arwen.AsyncContext{}
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetOriginalTxHash())
+	buff := storage.GetStorage(storageKey)
+	if len(buff) == 0 {
+		return asyncContext, nil
+	}
+
+	err := json.Unmarshal(buff, &asyncContext)
+	if err != nil {
+		return nil, err
+	}
+
+	return asyncContext, nil
+}
+
+func (host *vmHost) deleteCurrentAsyncContext() error {
+	runtime := host.Runtime()
+	storage := host.Storage()
+
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetOriginalTxHash())
```

Comment:
Yes, the original tx hash prevents reentrance, which is likely to be a useful feature. Will reconsider. Added TODO, and will fix on a new branch.

Reply:
The action was provoked by the previous tx hash, not the original hash. In case of a multi sig contract calling a smart contract which has several asyncs those asyncs should work with previous hash

Reply:
Replaced with `runtime.GetPrevTxHash()`. Added a few TODOs to ensure that `PrevTxHash` is different from `CurrentTxHash`, where required (between steps of an async call): otherwise, cross-shard async will differ from in-shard async, and contracts will not be able to rely on different tx hashes (e.g. `CurrentTxHash` in original SC call currently equals `CurrentTxHash` in `callBack()`, and must be not equal).